bin/migrate-oats-data/noi/noi_decisions/noi_components/notice_of_intent_decision_components_init.py

```diff
@@ -0,0 +1,199 @@
+from common import (
+    OATS_ETL_USER,
+    setup_and_get_logger,
+    add_timezone_and_keep_date_part,
+    ALRChangeCode,
+    OatsToAlcsAgCap,
+    OatsToAlcsAgCapSource,
+)
+from db import inject_conn_pool
+from psycopg2.extras import RealDictCursor, execute_batch
+
+etl_name = "init_notice_of_intent_decision_components"
+logger = setup_and_get_logger(etl_name)
+
+
+@inject_conn_pool
+def init_notice_of_intent_decision_components(conn=None, batch_size=1000):
+    """
+    This function is responsible for initializing the notice_of_intent_decision_components in ALCS.
+
+    Args:
+    conn (psycopg2.extensions.connection): PostgreSQL database connection. Provided by the decorator.
+    batch_size (int): The number of items to process at once. Defaults to BATCH_UPLOAD_SIZE.
+    """
+
+    logger.info(f"Start {etl_name}")
+    with conn.cursor(cursor_factory=RealDictCursor) as cursor:
+        with open(
+            "noi/sql/notice_of_intent_decisions/components/notice_of_intent_decision_components_insert_count.sql",
+            "r",
+            encoding="utf-8",
+        ) as sql_file:
+            count_query = sql_file.read()
+            cursor.execute(count_query)
+            count_total = dict(cursor.fetchone())["count"]
+        logger.info(
+            f"Total Notice of Intent Decision Components data to insert: {count_total}"
+        )
+
+        failed_inserts = 0
+        successful_inserts_count = 0
+        last_decision_component_id = 0
+
+        tmp_counter = 0
+
+        with open(
+            "noi/sql/notice_of_intent_decisions/components/notice_of_intent_decision_components_insert.sql",
+            "r",
+            encoding="utf-8",
+        ) as sql_file:
+            query = sql_file.read()
+
+            while True:
+                cursor.execute(
+                    f"""{query} 
+                    WHERE oaac.alr_appl_component_id > {last_decision_component_id} ORDER BY oaac.alr_appl_component_id;"""
```

Comment:
I pointed this out to @lstod and I can't remember what we decided, but doesn't this do a lot of duplicate work? It first selects all rows, but only takes the first batch, then reselects all the rows from the end of that batch, etc. Would it make sense to use a limit clause or add an upper bound of `last_decision_component_id + batch_size`?

Reply:
it does not actually read the queries from db, rows = cursor.fetchmany(batch_size) bellow it does the retrieving from DB

Reply:
My concern is that it needs to run the whole query on the database first, which does fetch all of the rows, but then only imports the batch size into Python. This isn't really the place for this discussion though (I probably shouldn't have brought it up), so I'm just going to mark it resolved. It might be worth looking into as a potential optimization.